File: lab2-num/lab2-num/java/Sqpolyroot.java
public class Sqpolyroot 
{

    public static void main(String[] args)
    {
/*
        double a = 1;
        double b = 2;
        double c = -1;
     
        double solution =
            (-b + Math.sqrt(Math.pow(b,2) - 4 * a * c))/(2*a);
        System.out.println(solution);
*/

/*      
        byte i = 120;

        the following will loop forever
        unless you *enable assertions* when executing: java -ea Sqpolyroot
       while ( i < 128 )
       {
            byte iBefore = i;
            i++;
            System.out.println("i = " + i);
            assert i > iBefore : "value in variable i has overflown";
        }
*/

        int i = 2147483647; // 2^31-1, the largest int value
        i = i + 2; // overflow
        System.out.println(i); // prints: -2147483647
        i = 2*i; // overflow
        System.out.println(i); // prints: 2

        double tiny = 1.0E-200;
        double huge = 1.0E+200;

        System.out.println("tiny = " + tiny);
        System.out.println("huge = " + huge);

        System.out.println("tiny * tiny = " + (tiny * tiny));
        System.out.println("huge * huge = " + (huge * huge));
        System.out.println("(huge == tiny + huge) = " + (huge == tiny + huge));

        System.out.println("huge * tiny * huge = " + huge * tiny * huge);
        System.out.println("huge * huge * tiny = " + huge * huge * tiny);
        System.out.println("(huge * huge) * (tiny * tiny) = " + (huge * huge) * (tiny * tiny));
        System.out.println("(huge * huge) - (huge * huge) = " + ((huge * huge) - (huge * huge)));
    }
}

File: lab2-num/haskell/sqpolyroot.hs
module Main where

import Data.Ratio

main = 
    do
--     print $ solution 1 2 (-1)
    printL "tiny*tiny" (tiny*tiny)
    printL "huge*huge" (huge*huge)
    printL "huge == tiny + huge" (huge == tiny + huge)
    printL "huge * tiny * huge" (huge * tiny * huge)
    printL "huge * huge * tiny" (huge * huge * tiny)
    printL "(huge * huge) * (tiny * tiny)" ((huge * huge) * (tiny * tiny))
    printL "(huge * huge) - (huge * huge)" ((huge * huge) - (huge * huge))

    
solution a b c = 
    (- b + (sqrt $ b ** 2 - 4 * a * c))/(2 * a)
    
    
tiny = 1.0E-200 :: Double
huge = 1.0E+200 :: Double

printL label x = do putStr (label ++ ": "); print x
